pkg/collector/distinct_value_collector.go

```diff
@@ -5,66 +5,83 @@ import (
 )
 
 type DistinctValue[T comparable] struct {
-	values   map[T]struct{}
-	new      map[T]struct{}
-	len      func(T) int
-	maxLen   int
-	currLen  int
-	totalLen int
-	mtx      sync.RWMutex
+	values      map[T]struct{}
+	new         map[T]struct{}
+	len         func(T) int
+	maxLen      int
+	currLen     int
+	limExceeded bool
+	diffEnabled bool
+	mtx         sync.Mutex
 }
 
 // NewDistinctValue with the given maximum data size. This is calculated
 // as the total length of the recorded strings. For ease of use, maximum=0
 // is interpreted as unlimited.
+// Use NewDistinctValueWithDiff to enable diff support, but that one is slightly slower.
 func NewDistinctValue[T comparable](maxDataSize int, len func(T) int) *DistinctValue[T] {
 	return &DistinctValue[T]{
-		values: make(map[T]struct{}),
-		new:    make(map[T]struct{}),
-		maxLen: maxDataSize,
-		len:    len,
+		values:      make(map[T]struct{}),
+		new:         make(map[T]struct{}),
+		maxLen:      maxDataSize,
+		diffEnabled: false, // disable diff to make it faster
+		len:         len,
 	}
 }
 
+// NewDistinctValueWithDiff is like NewDistinctValue but with diff support enabled.
+func NewDistinctValueWithDiff[T comparable](maxDataSize int, len func(T) int) *DistinctValue[T] {
+	return &DistinctValue[T]{
+		values:      make(map[T]struct{}),
+		new:         make(map[T]struct{}),
+		maxLen:      maxDataSize,
+		diffEnabled: true,
+		len:         len,
+	}
+}
+
+// Collect adds a new value to the distinct value collector.
+// return true when it reaches the limits and can't fit more values.
+// callers of return of Collect or call Exceeded to stop early.
 func (d *DistinctValue[T]) Collect(v T) (exceeded bool) {
-	d.mtx.RLock()
-	if _, ok := d.values[v]; ok {
-		d.mtx.RUnlock()
-		return // Already present
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+
+	if d.limExceeded {
+		return true
 	}
-	d.mtx.RUnlock()
 
 	// Calculate length
 	valueLen := d.len(v)
 
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
+	// Can it fit?
+	// note: we will stop adding values slightly before the limit is reached
+	if d.maxLen > 0 && d.currLen+valueLen >= d.maxLen {
+		// No, it can't fit
+		d.limExceeded = true
+		return true
+	}
 
 	if _, ok := d.values[v]; ok {
 		return // Already present
 	}
 
-	// Record total inspected length regardless
-	d.totalLen += valueLen
-
-	// Can it fit?
-	if d.maxLen > 0 && d.currLen+valueLen > d.maxLen {
-		// No
-		return true
+	if d.diffEnabled {
+		d.new[v] = struct{}{}
 	}
 
-	d.new[v] = struct{}{}
 	d.values[v] = struct{}{}
 	d.currLen += valueLen
+
 	return false
 }
 
 // Values returns the final list of distinct values collected and sorted.
 func (d *DistinctValue[T]) Values() []T {
 	ss := make([]T, 0, len(d.values))
```

Comment:
@joe-elliott I think we have the same subtle bug here as well, that we had in Diff.

we are checking the len of a member var without locking. I think we got lucky here because we usualy call Values after we are done collecting. 

locking this as well.